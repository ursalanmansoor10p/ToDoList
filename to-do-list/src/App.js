import "./App.css";
import React, { useEffect, useState } from "react";
import Title from "./components/Title";
import AddTodo from "./components/AddTodo";
import Todo from "./components/TodoDisplay";
import {
  collection,
  query,
  orderBy,
  onSnapshot,
  doc,
  updateDoc,
  deleteDoc,
  getDocs,
  where,
} from "firebase/firestore";
import { db } from "./firebase";
import { ImList } from "react-icons/im";
import { BsCalendarCheck } from "react-icons/bs";
import CalendarTask from "./components/CalendarTask";

function App() {
  const [todos, setTodos] = React.useState([]);
  const [td, setTd] = React.useState([]);
  const tdCollectionRef = collection(db, "todos");

  // when list view is true we will get the normal task
  // when false we will get completed calendar tasks
  // whole code is depended on list view. If true
  // left tab is open and displays both completed and todo
  // tasks. Otherwise it shows the calendar task with
  // completed tasks
  const [listView, selectListView] = useState(true);

  //Fetch the data from firebase db using useEffect Hook.
  //the [] as the second parameter means the useEffect will only run once
  React.useEffect(() => {
    //reference to the collection

    const q = query(collection(db, "todos"));
    console.log("q", q);
    // if you use the onSnapshot() method you constantly
    // listen to a document as oppose to get() which listens once
    // we want to listen constantly in order for app to update
    // in real time
    const unsub = onSnapshot(q, (querySnapshot) => {
      let todosArray = [];
      querySnapshot.forEach((doc) => {
        todosArray.push({ ...doc.data(), id: doc.id });
      });
      setTodos(todosArray);
    });
    return () => unsub();
  }, []);

  // toggle complete. The reference to this is passed down to
  // child component

  const toggleComplete = async (todo) => {
    await updateDoc(doc(db, "todos", todo.id), {
      completed: !todo.completed,
    });
  };
  // Delete functionality. The reference to this is passed down to
  // child component
  const handleDelete = async (id) => {
    await deleteDoc(doc(db, "todos", id));
  };

  console.log(todos);

  // Create a query against the collection
  console.log(listView);

  return (
    <div className="App">
      <div className="form_design">
        <div className="tabs">
          {/* if list view is true then take out borders of left
         button and grey out right icon */}
          <button
            className={listView ? "tab_btn left active" : "tab_btn left"}
            onClick={() => selectListView(true)}
          >
            <ImList className={listView ? "tab_icons" : "tab_icons disabled"} />
          </button>
          {/* if list view is false so the opposite*/}
          <button
            className={!listView ? "tab_btn right active" : "tab_btn right"}
            onClick={() => selectListView(false)}
          >
            <BsCalendarCheck
              className={listView ? "tab_icons disabled" : "tab_icons"}
            />
          </button>
        </div>
        {/* if listview is true then pass then display  */}

        {listView ? (
          <div>
            {todos.map((todo) => (
              // todo is db iteration passed down as object
              <div key={todo.id}>
                <Todo
                  data-testid="todo"
                  key={todo.id}
                  todo={todo}
                  toggleComplete={toggleComplete}
                  handleDelete={handleDelete}
                />
              </div>
            ))}
            <AddTodo />
          </div>
        ) : (
          // if list view is false show CalenderTask
          // todos is the db state
          <CalendarTask todos={todos} />
        )}
      </div>
    </div>
  );
}

export default App;
