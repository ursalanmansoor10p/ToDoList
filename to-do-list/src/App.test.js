import React from "react";
import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import App from "./App";
import { onSnapshot } from "firebase/firestore";

const todos = [
  { title: "abc", completed: false },
  { title: "xyz", completed: true },
];
const docs = [
  { id: 1, data: () => todos[0] },
  { id: 2, data: () => todos[1] },
];

jest.mock("firebase/firestore", () => ({
  collection: jest.fn(),
  addDoc: jest.fn(),
  query: jest.fn(),
  onSnapshot: jest.fn(),
  deleteDoc: jest.fn(),
}));
jest.mock("./firebase", () => ({
  db: jest.fn(),
}));

describe("App Component", () => {
  beforeEach(() => {
    onSnapshot.mockImplementation((_, callback) => {
      callback(docs);
      return jest.fn();
    });
  });
  test("should match snapshot", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  test("should fetch data from firebase on mount", () => {
    render(<App />);
    expect(onSnapshot).toBeCalled();
  });

  test("should display todos list by default", () => {
    render(<App />);
    const components = screen.getAllByTestId("todo-display");
    expect(components.length).toBe(todos.length);
    const addTodo = screen.getByTestId('add-todo');
    expect(addTodo).toBeInTheDocument();
  });

  test("should toggle between todos list and calendar task when clicking on the tabs", () => {
    render(<App />);
    const [leftTab, rightTab] = screen.getAllByRole("button");
    fireEvent.click(rightTab);
    const calendar = screen.getByTestId("calendar-tasks");
    expect(calendar).toBeInTheDocument();
    fireEvent.click(leftTab);
    const todoDisplay = screen.getAllByTestId("todo-display");
    expect(todoDisplay.length).toBe(todos.length);
  });
});
