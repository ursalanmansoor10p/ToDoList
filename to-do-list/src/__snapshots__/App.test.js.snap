// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should match snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="form_design"
    >
      <div
        class="tabs"
      >
        <button
          class="tab_btn left active"
        >
          <svg
            class="tab_icons"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            version="1.1"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h4v4h-4zM6 1h10v2h-10zM0 6h4v4h-4zM6 7h10v2h-10zM0 12h4v4h-4zM6 13h10v2h-10z"
            />
          </svg>
        </button>
        <button
          class="tab_btn right"
        >
          <svg
            class="tab_icons disabled"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.854 7.146a.5.5 0 0 1 0 .708l-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 1 1 .708-.708L7.5 9.793l2.646-2.647a.5.5 0 0 1 .708 0z"
            />
            <path
              d="M3.5 0a.5.5 0 0 1 .5.5V1h8V.5a.5.5 0 0 1 1 0V1h1a2 2 0 0 1 2 2v11a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2V3a2 2 0 0 1 2-2h1V.5a.5.5 0 0 1 .5-.5zM1 4v10a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V4H1z"
            />
          </svg>
        </button>
      </div>
      <div>
        <div>
          <div
            class="to"
            data-testid="todo-display"
          >
            <label
              class=""
              style="display: flex; align-items: center;"
            >
              <span>
                <div
                  class=""
                  style="height: 18px; width: 18px; border-width: 2px; border-color: rgba(0, 0, 0, 0.452); border-style: solid; border-radius: 5px; display: flex; align-items: center; justify-content: center;"
                >
                  <input
                    data-testid="checkbox"
                    hidden=""
                    name=""
                    type="checkbox"
                    value=""
                  />
                </div>
              </span>
            </label>
            <div
              data-testid="todo-title"
            >
              abc
            </div>
            <div>
              <div>
                <button
                  class="button-delete"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="to"
            data-testid="todo-display"
          >
            <label
              class=""
              style="display: flex; align-items: center;"
            >
              <span>
                <div
                  class=""
                  style="height: 18px; width: 18px; border-width: 2px; border-color: rgba(0, 0, 0, 0.452); border-style: solid; border-radius: 5px; display: flex; align-items: center; justify-content: center;"
                >
                  <img
                    alt=""
                    src="check.png"
                    style="width: 24px;"
                  />
                  <input
                    checked=""
                    data-testid="checkbox"
                    hidden=""
                    name=""
                    type="checkbox"
                    value=""
                  />
                </div>
              </span>
            </label>
            <div
              data-testid="todo-title"
            >
              <del>
                xyz
              </del>
            </div>
            <div>
              <div>
                <button
                  class="button-delete"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          data-testid="add-todo"
        >
          <form
            name="add-form"
          >
            <div
              class="input_container"
            >
              <input
                class="input_style"
                placeholder="Please enter text"
                type="text"
                value=""
              />
              <div
                class="btn_container"
              >
                <button
                  class="button button1"
                >
                  +
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
